# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [11]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest

from scipy.stats import ttest_ind, norm, t
from scipy.stats import f_oneway
from scipy.stats import sem

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [12]:
# LOAD SALARY DATA

# PSEUDOCODE:
# - Go up one folder from your-code
# - Enter the data folder
# - Load the salary CSV into a DataFrame called salaries

salaries = pd.read_csv(
    "../data/Current_Employee_Names__Salaries__and_Position_Titles.csv"
)

In [13]:
salaries.head()

,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86


Examine the `salaries` dataset using the `head` function below.

In [14]:
# CHECK SALARY DATA STRUCTURE

# PSEUDOCODE:
# - Inspect column names
# - Check data types
# - Check for missing values

salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33183 entries, 0 to 33182
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               33183 non-null  object 
 1   Job Titles         33183 non-null  object 
 2   Department         33183 non-null  object 
 3   Full or Part-Time  33183 non-null  object 
 4   Salary or Hourly   33183 non-null  object 
 5   Typical Hours      8022 non-null   float64
 6   Annual Salary      25161 non-null  float64
 7   Hourly Rate        8022 non-null   float64
dtypes: float64(3), object(5)
memory usage: 2.0+ MB


In [15]:
salaries.columns

Index(['Name', 'Job Titles', 'Department', 'Full or Part-Time',
       'Salary or Hourly', 'Typical Hours', 'Annual Salary', 'Hourly Rate'],
      dtype='object')

In [16]:
# CHECK EMPLOYEE PAYMENT TYPES

salaries["Salary or Hourly"].value_counts()

Salary or Hourly
Salary    25161
Hourly     8022
Name: count, dtype: int64

In [17]:
# SELECT HOURLY WORKERS

# PSEUDOCODE:
# - Keep only employees where Salary or Hourly equals "Hourly"
# - Select their Hourly Rate values
# - Remove missing values

hourly_workers = salaries[
    salaries["Salary or Hourly"] == "Hourly"
]

hourly_wages = hourly_workers["Hourly Rate"].dropna()

hourly_wages.head()

4     19.86
6     46.10
7     35.60
10     2.65
18    17.68
Name: Hourly Rate, dtype: float64

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [18]:
# CHALLENGE 2 - ONE-SAMPLE HYPOTHESIS TEST

# PSEUDOCODE:
# - Null hypothesis H0: mean hourly wage = 30
# - Alternative hypothesis H1: mean hourly wage != 30
# - Use a two-sided one-sample t-test
# - Significance level alpha = 0.05
# - Compare the p-value with alpha

from scipy.stats import ttest_1samp

alpha = 0.05
hypothesized_mean = 30

t_statistic, p_value = ttest_1samp(
    hourly_wages,
    popmean=hypothesized_mean
)

print("t-statistic:", t_statistic)
print("p-value:", p_value)

if p_value < alpha:
    print("Reject H0: the mean hourly wage is significantly different from $30.")
else:
    print("Fail to reject H0: there is not enough evidence that the mean hourly wage differs from $30.")

t-statistic: 20.6198057854942
p-value: 4.3230240486229894e-92
Reject H0: the mean hourly wage is significantly different from $30.


# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [19]:
# CHALLENGE 3 - 95% CONFIDENCE INTERVAL

# PSEUDOCODE:
# - Use the hourly wage sample
# - Calculate the sample mean
# - Calculate the standard error
# - Use a t-distribution because population standard deviation is unknown
# - Use 95% confidence level
# - Degrees of freedom = n - 1

confidence_level = 0.95

sample_mean = hourly_wages.mean()
standard_error = sem(hourly_wages)
degrees_of_freedom = len(hourly_wages) - 1

confidence_interval = t.interval(
    confidence_level,
    df=degrees_of_freedom,
    loc=sample_mean,
    scale=standard_error
)

print("Sample mean:", sample_mean)
print("95% confidence interval:", confidence_interval)

Sample mean: 32.78855771628024
95% confidence interval: (32.52345834488425, 33.05365708767623)


The sample mean hourly wage is approximately $32.79, with a 95% confidence interval from $32.52 to $33.05. Since $30 is not included in this interval, this result is consistent with rejecting the null hypothesis that the true mean hourly wage is $30.

# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [20]:
# CHALLENGE 4 - ONE-SAMPLE PROPORTION TEST

# PSEUDOCODE:
# - H0: proportion of hourly workers = 25%
# - H1: proportion of hourly workers != 25%
# - Count how many employees are hourly workers
# - Count the total number of employees
# - Perform a two-sided one-sample proportion z-test
# - Compare the p-value with alpha = 0.05

from statsmodels.stats.proportion import proportions_ztest

alpha = 0.05
hypothesized_proportion = 0.25

# Count hourly workers
hourly_count = (
    salaries["Salary or Hourly"] == "Hourly"
).sum()

# Total number of employees
total_count = len(salaries)

# Sample proportion
sample_proportion = hourly_count / total_count

# Run the proportion z-test
z_statistic, p_value = proportions_ztest(
    count=hourly_count,
    nobs=total_count,
    value=hypothesized_proportion,
    alternative="two-sided"
)

print("Number of hourly workers:", hourly_count)
print("Total employees:", total_count)
print("Sample proportion:", sample_proportion)
print("z-statistic:", z_statistic)
print("p-value:", p_value)

if p_value < alpha:
    print(
        "Reject H0: the proportion of hourly workers "
        "is significantly different from 25%."
    )
else:
    print(
        "Fail to reject H0: there is not enough evidence "
        "that the proportion of hourly workers differs from 25%."
    )

Number of hourly workers: 8022
Total employees: 33183
Sample proportion: 0.24175029382515142
z-statistic: -3.5099964213703005
p-value: 0.00044811272490579676
Reject H0: the proportion of hourly workers is significantly different from 25%.


The sample shows that about 24.18% of employees are hourly workers. The p-value is much smaller than 0.05, so we reject the null hypothesis. This means the proportion of hourly workers is statistically significantly different from 25%.